src/transformers/modeling_utils.py

```diff
@@ -1282,6 +1292,18 @@ def load(module: nn.Module, prefix=""):
 
         return model
 
+    def retrieve_modules_from_weights(self, weight_keys):
+        retrieved_modules = set()
+        for weight_key in weight_keys:
```

Comment:
Can it be done with using `self.named_modules` instead? I was thinking of something like
```
module_keys = [".".join(key.split(".")[:-1]) for key in weight_keys]
return [module for name, module in self.named_modules() if name in module_keys]
```

It would be cleaner.

Reply:
Both implementations proposed overlook the weights of the base transformer model.

This is an issue as if some weights of the base transformer model are missing in the checkpoint file, they don't end up in the `unintialized_modules` returned by this method; they're not used with the `init_weight()` method.

See below for a reproducible code example. If one saves a torch state dict without the base model prefix and reloads it in a model that has the model prefix, then the two methods here are unable to identify the missing weights. Such a use-case is for dropping a model head, for example.

```py
from transformers import BertModel, BertForSequenceClassification, BertConfig
import torch

model = BertModel.from_pretrained("bert-base-uncased")
state_dict = model.state_dict()
del state_dict["embeddings.word_embeddings.weight"]

model.save_pretrained("save_it_here")
torch.save(state_dict, "save_it_here/pytorch_model.bin")

model = BertForSequenceClassification.from_pretrained("save_it_here", initializer_range=0.00001)

print(model.bert.embeddings.word_embeddings.weight)
```

This crashes with the current implementation, and with @sgugger's proposal, the resulting embeddings have the following values on my machine (way above what they should be given the initializer range given above):
```
tensor([[ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.5985,  0.1010,  0.0665,  ..., -0.6634, -0.4788, -0.5904],
        [ 3.4545,  0.4866,  0.8241,  ..., -1.4983, -0.4970, -1.3458],
```

on `master` I get the following output:

```
tensor([[ 0.0000e+00,  0.0000e+00,  0.0000e+00,  ...,  0.0000e+00,
          0.0000e+00,  0.0000e+00],
        [ 6.3332e-06,  5.1651e-07, -3.1017e-06,  ...,  4.8352e-07,
          1.2032e-05,  5.6933e-06],
```

This is a bit of a far-fetched use-case as I can't imagine when one would delete a value from the state dict this way, it's just to get an idea of something we're missing.

Reply:
One of the suggestions I made is that we have a kill switch for odd cases, i.e. init_all_weights, so 
- in the general case things will be fast
- in the edge cases correctness should be possible

I guess the only difficulty here is that un-inited weights could go unnoticed and lead to a subtle problems that would be hard to diagnose.

And that's why another proposal was very simply - just leave everything as it is now and give a user a way to turn the optimized loading if they want to. e.g. I have been commenting out this code for months now, so I know what I'm doing, so I just want a flag that I can tell `from_pretrained` to skip the weights init.

But now we have a much more refined solution, which is preferrable to all or nothing. But still, if we are worried about edge cases and correctness is super important, I suggest that we implement Patrick's way, but have an explicit flag to turn it on.